From 99cee6603f92dfd67e5003899b90b28cfb2274b2 Mon Sep 17 00:00:00 2001
From: Caio Felipe Cruz <caio.soaresdacruz@windriver.com>
Date: Fri, 3 Oct 2025 11:07:44 -0400
Subject: [PATCH] ice: fix DPLL pin handling for SMA/UFL pin types

Add support in ice_dpll_find_pin() for ICE_DPLL_PIN_TYPE_SOFTWARE to
correctly map SMA and UFL software pins. Additionally, ensure DPLL pin
updates are protected by ice_dpll_pin_update_lock()/unlock().

Signed-off-by: Caio Felipe Cruz <caio.soaresdacruz@windriver.com>
---
 src/ice_dpll.c | 13 +++++++++++++
 src/ice_ptp.c  |  2 ++
 2 files changed, 15 insertions(+)

diff --git a/src/ice_dpll.c b/src/ice_dpll.c
index e27118b..012096b 100644
--- a/src/ice_dpll.c
+++ b/src/ice_dpll.c
@@ -3761,6 +3761,19 @@ ice_dpll_find_pin(struct ice_pf *pf, enum ice_dpll_pin_type pin_type,
 		if (pin_idx >= pf->dplls.num_outputs)
 			return NULL;
 		return &pf->dplls.outputs[pin_idx];
+	case ICE_DPLL_PIN_TYPE_SOFTWARE:
+		switch (pin_idx) {
+		case SMA_SMA1:
+			return &pf->dplls.sma[ICE_DPLL_PIN_SW_1_IDX];
+		case SMA_UFL1:
+			return &pf->dplls.ufl[ICE_DPLL_PIN_SW_1_IDX];
+		case SMA_SMA2:
+			return &pf->dplls.sma[ICE_DPLL_PIN_SW_2_IDX];
+		case SMA_UFL2:
+			return &pf->dplls.ufl[ICE_DPLL_PIN_SW_2_IDX];
+		default:
+			return NULL;
+		}
 	default:
 		return NULL;
 	}
diff --git a/src/ice_ptp.c b/src/ice_ptp.c
index 1f3c4d9..a0ec958 100644
--- a/src/ice_ptp.c
+++ b/src/ice_ptp.c
@@ -380,7 +380,9 @@ static ssize_t ice_ptp_sma_cfg_store(struct device *dev, const char *buf,
 		return err;
 
 	pf->hw.ptp.sma_cfg[pin] = fun;
+	ice_dpll_pin_update_lock(pf);
 	ice_ptp_set_sma_cfg(&pf->hw);
+	ice_dpll_pin_update_unlock(pf, true, ICE_DPLL_PIN_TYPE_SOFTWARE, pin);
 
 	return len;
 
-- 
2.34.1

